pandas/core/arrays/datetimes.py

```diff
@@ -125,17 +129,31 @@ def wrapper(self, other):
                 return ops.invalid_comparison(self, other, op)
 
             if is_object_dtype(other):
-                result = op(self.astype('O'), np.array(other))
+                # messy... Previously, DatetimeArray.astype(object) -> Index
+                # now it's an ndarray. op[ndarray, ndarray] doesn't
+                # doesn't raise when comparing tz and non-tz (just returns
+                # False.
+                with np.errstate(all='ignore'):
+                    result = ops._comp_method_OBJECT_ARRAY(op,
+                                                           self.astype(object),
+                                                           other)
                 o_mask = isna(other)
             elif not (is_datetime64_dtype(other) or
                       is_datetime64tz_dtype(other)):
                 # e.g. is_timedelta64_dtype(other)
                 return ops.invalid_comparison(self, other, op)
             else:
                 self._assert_tzawareness_compat(other)
-                if not hasattr(other, 'asi8'):
-                    # ndarray, Series
-                    other = type(self)(other)
+                if isinstance(other, (ABCIndexClass, ABCSeries)):
+                    other = other.array
+
+                if (is_datetime64_dtype(other) and
+                        not is_datetime64_ns_dtype(other) or
+                        not hasattr(other, 'asi8')):
```

Comment:
why is the asi8 check needed?

Reply:
Else a datetime64[ns] ndarray would not be boxed in a DatetimeArray, and things like `other._isnan` would fail.